### PR TITLE
feat(backend): build and cache program_restrictions

### PR DESCRIPTION
**Pull request by imagine-hussain**
## Features
- Builder for the conditions in `runprocessers` that will cache the `ProgramRestrictions`
- `GenEd` Category (unimplemented - simply does no match for now; out of scope)
- Fixed silent bug from using `[]` as default value instead of `{}` that mypy didnt catch (thanks mypy)

## Hotfixes:
- I can't spell and had some errors because of inconsistent use of `dependency` and `dependancy`; settles on unifying everything to use `dependency`
- Missed that a general education category can also exist as a dependent; found this out when trying to create all Restrictions

### CODE REVIEW

**Comment by imagine-hussain** [comment]
# Created Conditions:


```
3134:  CompositeRestriction: Logic::(Logic.AND) Restrictions::(['MaturityCondition: Dependency: {"UOC": 24, "category": "all courses"}, self.dependent: level 2 courses', 'MaturityCondition: Dependency: {"UOC": 48, "category": "all courses"}, self.dependent: level 3 courses'])
3462:  CompositeRestriction: Logic::(Logic.AND) Restrictions::(['MaturityCondition: Dependency: {"UOC": 24, "category": "all courses"}, self.dependent: level 2 courses', 'MaturityCondition: Dependency: {"UOC": 48, "category": "all courses"}, self.dependent: level 3 courses'])
3502:  CompositeRestriction: Logic::(Logic.AND) Restrictions::(['MaturityCondition: Dependency: {"UOC": 24, "category": "all courses"}, self.dependent: level 2 courses', 'MaturityCondition: Dependency: {"UOC": 48, "category": "all courses"}, self.dependent: level 3 courses'])
3523:  CompositeRestriction: Logic::(Logic.AND) Restrictions::(['MaturityCondition: Dependency: {"UOC": 24, "category": "all courses"}, self.dependent: level 2 courses', 'MaturityCondition: Dependency: {"UOC": 48, "category": "all courses"}, self.dependent: level 3 courses'])
3529:  CompositeRestriction: Logic::(Logic.AND) Restrictions::(['MaturityCondition: Dependency: {"UOC": 24, "category": "all courses"}, self.dependent: level 2 courses', 'MaturityCondition: Dependency: {"UOC": 48, "category": "all courses"}, self.dependent: level 3 courses'])
3543:  CompositeRestriction: Logic::(Logic.AND) Restrictions::(['MaturityCondition: Dependency: {"cores": null, "category": "(level 1 courses && ECON courses)"}, self.dependent: (level 3 courses && ECON courses)'])
3554:  CompositeRestriction: Logic::(Logic.AND) Restrictions::(['MaturityCondition: Dependency: {"UOC": 24, "category": "all courses"}, self.dependent: level 2 courses', 'MaturityCondition: Dependency: {"UOC": 72, "category": "all courses"}, self.dependent: level 3 courses', 'MaturityCondition: Dependency: {"UOC": 120, "category": "all courses"}, self.dependent: level 4 courses'])
3558:  CompositeRestriction: Logic::(Logic.AND) Restrictions::(['MaturityCondition: Dependency: {"UOC": 24, "category": "all courses"}, self.dependent: level 2 courses', 'MaturityCondition: Dependency: {"UOC": 48, "category": "all courses"}, self.dependent: level 3 courses'])
3563:  CompositeRestriction: Logic::(Logic.AND) Restrictions::(['MaturityCondition: Dependency: {"cores": null, "category": "(level 1 courses && ECON courses)"}, self.dependent: (level 3 courses && ECON courses)'])
3564:  CompositeRestriction: Logic::(Logic.AND) Restrictions::(['MaturityCondition: Dependency: {"cores": null, "category": "(level 1 courses && ECON courses)"}, self.dependent: (level 3 courses && ECON courses)'])
3566:  CompositeRestriction: Logic::(Logic.AND) Restrictions::(['MaturityCondition: Dependency: {"cores": null, "category": "(level 1 courses && ECON courses)"}, self.dependent: (level 3 courses && ECON courses)'])
3573:  CompositeRestriction: Logic::(Logic.AND) Restrictions::(['MaturityCondition: Dependency: {"UOC": 24, "category": "all courses"}, self.dependent: level 2 courses', 'MaturityCondition: Dependency: {"UOC": 48, "category": "all courses"}, self.dependent: level 3 courses'])
3574:  CompositeRestriction: Logic::(Logic.AND) Restrictions::(['MaturityCondition: Dependency: {"cores": null, "category": "(level 1 courses && ECON courses)"}, self.dependent: (level 3 courses && ECON courses)'])
3584:  CompositeRestriction: Logic::(Logic.AND) Restrictions::(['MaturityCondition: Dependency: {"UOC": 24, "category": "all courses"}, self.dependent: level 2 courses', 'MaturityCondition: Dependency: {"UOC": 48, "category": "all courses"}, self.dependent: level 3 courses'])
3587:  CompositeRestriction: Logic::(Logic.AND) Restrictions::(['MaturityCondition: Dependency: {"UOC": 24, "category": "all courses"}, self.dependent: level 2 courses', 'MaturityCondition: Dependency: {"UOC": 48, "category": "all courses"}, self.dependent: level 3 courses'])
3593:  CompositeRestriction: Logic::(Logic.AND) Restrictions::(['MaturityCondition: Dependency: {"UOC": 24, "category": "all courses"}, self.dependent: level 2 courses', 'MaturityCondition: Dependency: {"UOC": 48, "category": "all courses"}, self.dependent: level 3 courses'])
3597:  CompositeRestriction: Logic::(Logic.AND) Restrictions::(['MaturityCondition: Dependency: {"UOC": 24, "category": "all courses"}, self.dependent: level 2 courses', 'MaturityCondition: Dependency: {"UOC": 48, "category": "all courses"}, self.dependent: level 3 courses'])
3598:  CompositeRestriction: Logic::(Logic.AND) Restrictions::(['MaturityCondition: Dependency: {"UOC": 24, "category": "all courses"}, self.dependent: level 2 courses', 'MaturityCondition: Dependency: {"UOC": 48, "category": "all courses"}, self.dependent: level 3 courses'])
3599:  CompositeRestriction: Logic::(Logic.AND) Restrictions::(['MaturityCondition: Dependency: {"UOC": 24, "category": "all courses"}, self.dependent: level 2 courses', 'MaturityCondition: Dependency: {"UOC": 48, "category": "all courses"}, self.dependent: level 3 courses'])
3673:  CompositeRestriction: Logic::(Logic.AND) Restrictions::(['MaturityCondition: Dependency: {"cores": null, "category": "(level 1 courses && ECON courses)"}, self.dependent: (level 3 courses && ECON courses)'])
3734:  CompositeRestriction: Logic::(Logic.AND) Restrictions::(['MaturityCondition: Dependency: {"UOC": 24, "category": "all courses"}, self.dependent: level 2 courses', 'MaturityCondition: Dependency: {"UOC": 48, "category": "all courses"}, self.dependent: level 3 courses'])
3764:  CompositeRestriction: Logic::(Logic.AND) Restrictions::(['MaturityCondition: Dependency: {"UOC": 24, "category": "all courses"}, self.dependent: level 2 courses', 'MaturityCondition: Dependency: {"UOC": 48, "category": "all courses"}, self.dependent: level 3 courses'])
3784:  CompositeRestriction: Logic::(Logic.AND) Restrictions::(['MaturityCondition: Dependency: {"UOC": 24, "category": "all courses"}, self.dependent: level 2 courses', 'MaturityCondition: Dependency: {"UOC": 48, "category": "all courses"}, self.dependent: level 3 courses'])
3835:  CompositeRestriction: Logic::(Logic.AND) Restrictions::(['MaturityCondition: Dependency: {"UOC": 24, "category": "all courses"}, self.dependent: level 2 courses', 'MaturityCondition: Dependency: {"UOC": 48, "category": "all courses"}, self.dependent: level 3 courses'])
3964:  CompositeRestriction: Logic::(Logic.AND) Restrictions::(['MaturityCondition: Dependency: {"UOC": 48, "category": "all courses"}, self.dependent: level 3 courses'])
3980:  CompositeRestriction: Logic::(Logic.AND) Restrictions::(['MaturityCondition: Dependency: {"UOC": 48, "category": "all courses"}, self.dependent: GenEdCategory'])
4058:  CompositeRestriction: Logic::(Logic.AND) Restrictions::(['MaturityCondition: Dependency: {"cores": null, "category": "(level 1 courses && ECON courses)"}, self.dependent: (level 3 courses && ECON courses)'])
4733:  CompositeRestriction: Logic::(Logic.AND) Restrictions::(['MaturityCondition: Dependency: {"UOC": 24, "category": "all courses"}, self.dependent: level 2 courses', 'MaturityCondition: Dependency: {"UOC": 48, "category": "all courses"}, self.dependent: level 3 courses'])
4744:  CompositeRestriction: Logic::(Logic.AND) Restrictions::(['MaturityCondition: Dependency: {"cores": null, "category": "(level 1 courses && ECON courses)"}, self.dependent: (level 3 courses && ECON courses)'])
```

**Comment by imagine-hussain** [comment]
Everything not listed in the above created conditions, is of form

```
CODE: CompositeRestriction: Logic::(Logic.AND) Restrictions::([])```

**Comment by imagine-hussain** [comment]
Once this is done, the MaturityConditions is officially done.
Next, this just needs to be attached somewhere it can get evaluated.
Then, these same objects can be used for any other type of conditions that may be found

File: backend/algorithms/objects/categories.py [hunk 1]
```diff
@@ -149,3 +149,24 @@ def match_definition(self, course: str) -> bool:
 
     def __str__(self) -> str:
         return self.faculty
+
+
+class GenEdCategory(Category):
+    """
+    Category for a course being a Gened
+        - This is condition is unique in the sense that it's matching is
```
Comment (Peedee2002) [inline backend/algorithms/objects/categories.py:157]: Oh god does that mean we need to pass down gen eds?

**Review by Peedee2002 — APPROVED** [review]
cool
I didnt even notice the "dependent" misspelling

File: backend/algorithms/objects/categories.py (hunk 1 above)
Comment (imagine-hussain) [reply backend/algorithms/objects/categories.py:157]: Indeed; Either they get attached to the class or, the `is_allowed` method needs to do a lookup to something else.

An alternative is to pass in the `User` class to `match_definition` and then have some method on the user to list / match against gen-eds

**Pull request merged**